File: SortVision/public/code/selection/lua/selectionSort.lua
-- Selection Sort (in-place)
-- Time: Best/Avg/Worst O(n^2), Space: O(1)
local function selectionSort(arr)
  for i = 1, #arr - 1 do
    local minIdx = i
    for j = i + 1, #arr do
      if arr[j] < arr[minIdx] then
        minIdx = j
      end
    end
    if minIdx ~= i then
      arr[i], arr[minIdx] = arr[minIdx], arr[i]
    end
  end
  return arr
end

local arr = {64, 25, 12, 22, 11}
print(table.concat(selectionSort(arr), ", "))

File: SortVision/public/code/selection/c/selectionSort.c
#include <stdio.h>

/* Selection Sort (in-place)
   Time: Best/Avg/Worst O(n^2)
   Space: O(1) */
void selectionSort(int arr[], int n) {
    for (int i = 0; i < n - 1; i++) {
        int minIdx = i;
        for (int j = i + 1; j < n; j++) {
            if (arr[j] < arr[minIdx]) minIdx = j;
        }
        if (minIdx != i) {
            int t = arr[i];
            arr[i] = arr[minIdx];
            arr[minIdx] = t;
        }
    }
}

int main(void) {
    int arr[] = {64, 25, 12, 22, 11};
    int n = (int)(sizeof(arr) / sizeof(arr[0]));

    selectionSort(arr, n);
    for (int i = 0; i < n; i++) printf("%d ", arr[i]);
    printf("\n");
    return 0;
}
